rust/connlib/tunnel/src/client.rs

```diff
@@ -593,43 +600,46 @@ impl ClientState {
 
         self.gateways_site.insert(gateway_id, site_id);
         self.recently_connected_gateways.put(gateway_id, ());
+        self.peers
+            .add_ips_with_resource(&gateway_id, &ips, &resource_id);
 
-        if self.peers.get(&gateway_id).is_some() {
-            self.peers
-                .add_ips_with_resource(&gateway_id, &ips, &resource_id);
+        if !self.node.accepts_packet(gateway_id) {
+            let (offer, buffer) = self.node.new_connection(
+                gateway_id,
+                awaiting_connection_details.last_intent_sent_at,
+                now,
+            );
+            buffer.extend(awaiting_connection_details.packets);
 
-            return Ok(Some(Request::ReuseConnection(ReuseConnection {
+            return Ok(Request::NewConnection(RequestConnection {
                 resource_id,
                 gateway_id,
-                payload: awaiting_connection_details.domain,
-            })));
-        };
-
-        self.peers.insert(
-            GatewayOnClient::new(gateway_id, &ips, HashSet::from([resource_id])),
-            &[],
-        );
-        self.peers
-            .add_ips_with_resource(&gateway_id, &ips, &resource_id);
+                client_preshared_key: Secret::new(Key(*offer.session_key.expose_secret())),
+                client_payload: ClientPayload {
+                    ice_parameters: Offer {
+                        username: offer.credentials.username,
+                        password: offer.credentials.password,
+                    },
+                    domain: awaiting_connection_details.domain,
+                },
+            }));
+        }
 
-        let offer = self.node.new_connection(
-            gateway_id,
-            awaiting_connection_details.last_intent_sent_at,
-            Instant::now(),
-        );
+        for packet in awaiting_connection_details.packets {
+            match self.node.encapsulate(gateway_id, packet, now) {
+                Ok(Some(transmit)) => self.buffered_transmits.push_back(transmit.into_owned()),
+                Err(other) => {
+                    tracing::debug!("Failed to encapsulate buffered packet: {other}");
+                }
+                Ok(None) => {}
+            }
+        }
 
-        return Ok(Some(Request::NewConnection(RequestConnection {
+        return Ok(Request::ReuseConnection(ReuseConnection {
```

Comment:
Continuing from https://github.com/firezone/firezone/pull/6403#issuecomment-2305917354.

This is the only return-path of `ReuseConnection`. In order for this to race an in-flight `NewConnection` message, we would have to send a connection intent for a **different** resource than the one that is currently being established.

So at the very least, the `NewConnection` request has a head-start of the time difference between the packet that triggered the new connection and the packet that wants to talk to the other resource. I would assume that this bias is enough to not have this race condition occur in most cases?

Should we buffer `ReuseConnection` requests until we successfully setup the connection?

Reply:
Ah, this is a good catch. I would assume we probably should handle this gracefully if possible. Here's a real-world scenario:

- To successfully route GitHub traffic, I've added multiple DNS resources for Github (*.github.com, *.githubassets.com, etc)
- When loading Github.com in a webpage, multiple packets will be sent at the same time (within milliseconds of each other) by the browser

This will be fairly common if it isn't already the case for some customers.

Reply:
It is handled gracefully as part of https://github.com/firezone/firezone/pull/6403. The question is whether we need a compatibility gate here based on the versions but it appears that we do.

Reply:
yeah, seconding what @jamilbk is saying, the interface might have at the same time packets for multiple resources on the same gateway that remain unread, so they will be basically emitted with the smallest possible delay.

Reply:
Okay, a version gate it is then :)

Should we wait for a 1.3.0 release with this then? The gateway changes can go in at any point.

Reply:
Yeah this could be slated for 1.3.0 since we now know it's not really the reason behind the connection reset on macOS. Would the gateway require 1.3.0 too? Does this require any portal changes or we just need to make sure the Client and Gateway changes in this PR select each other?

Reply:
> Yeah this could be slated for 1.3.0 since we now know it's not really the reason behind the connection reset on macOS. Would the gateway require 1.3.0 too? Does this require any portal changes or we just need to make sure the Client and Gateway changes in this PR select each other?

#6403 contains some portal changes already. Other than that, nothing is needed.

Reply:
Ok just so I'm clear, that means the portal needs to be deployed _before_ this code is shipped in a Client, correct?